build-tools/src/main/java/org/elasticsearch/gradle/plugin/PluginBuildPlugin.java

```diff
@@ -114,64 +116,41 @@ private static void configureDependencies(final Project project) {
-                // intentionally an Action and not a lambda to avoid issues with up-to-date check
-                task.doLast(new Action<Task>() {
-                    @Override
-                    public void execute(Task task) {
-                        InputStream resourceTemplate = PluginBuildPlugin.class.getResourceAsStream("/" + templateFile.getName());
-                        try {
-                            String templateText = IOUtils.toString(resourceTemplate, StandardCharsets.UTF_8.name());
-                            FileUtils.write(templateFile, templateText, "UTF-8");
-                        } catch (IOException e) {
-                            throw new GradleException("Unable to copy plugin properties", e);
-                        }
-                    }
-                });
-            }
-        });
-
-        final var buildProperties = project.getTasks().register("pluginProperties", Copy.class, copy -> {
-            copy.dependsOn(copyPluginPropertiesTemplate);
-            copy.from(templateFile);
-            copy.into(new File(project.getBuildDir(), "generated-resources"));
-
-            if (extension.getName() == null) {
-                throw new InvalidUserDataException("name is a required setting for esplugin");
-            }
-
-            if (extension.getDescription() == null) {
-                throw new InvalidUserDataException("description is a required setting for esplugin");
-            }
-
-            if (extension.getType().equals(PluginType.BOOTSTRAP) == false && extension.getClassname() == null) {
-                throw new InvalidUserDataException("classname is a required setting for esplugin");
-            }
 
-            Map<String, Object> map = new LinkedHashMap<>(12);
-            map.put("name", extension.getName());
-            map.put("description", extension.getDescription());
-            map.put("version", extension.getVersion());
-            map.put("elasticsearchVersion", Version.fromString(VersionProperties.getElasticsearch()).toString());
-            map.put("javaVersion", project.getExtensions().getByType(JavaPluginExtension.class).getTargetCompatibility().toString());
-            map.put("classname", extension.getType().equals(PluginType.BOOTSTRAP) ? "" : extension.getClassname());
-            map.put("extendedPlugins", extension.getExtendedPlugins().stream().collect(Collectors.joining(",")));
-            map.put("hasNativeController", extension.isHasNativeController());
-            map.put("requiresKeystore", extension.isRequiresKeystore());
-            map.put("type", extension.getType().toString());
-            map.put("javaOpts", extension.getJavaOpts());
-            map.put("licensed", extension.isLicensed());
-
-            copy.expand(map);
-            copy.getInputs().properties(map);
+        final var buildProperties = project.getTasks().register("pluginProperties", GeneratePluginPropertiesTask.class, task -> {
+            /*task.doFirst(t -> {
```

Comment:
putting this in a doFirst block changes the behaviour here in a way that we fail later than before. If we keep the check in the configuration block we would fail when the task is materialized which is usually during task graph calculation. IMO we should just keep that behaviour and fail as early as we reliable can to shorten the feedback cycle for the user.

Reply:
If I can do those checks at that time, why do I need any of the lazy evaluation. Shouldn’t the entire extension be fully configured then?

Reply:
Are these explicit checks even  necessary? If these are `@Input`s on the task it should fail if they are `null`.

Reply:
I think it may be confusing getting the error message there, since a developer does not set this on the task, they set it on the extension.